### PR TITLE
Feature/add services

### PR DESCRIPTION
**Pull request by DmitriyG0r4anin**

### CODE REVIEW

**Review by MaliyOna — APPROVED** [review]

File: User.BusinessLogic/Services/Implementations/ClientService.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        var cache = await distributedCache.GetDataFromCacheAsync<ClientModel>(key, cancellationToken);
+
+        if (cache is not null)
+            return cache;
+
+        var client = await clientRepository.GetByIdAsync(id, cancellationToken);
+
+        var clientModel = client.Adapt<ClientModel>();
+
+        var cacheLifetime = TimeSpan.FromMinutes(10);
```
Comment (5Ka-me) [inline User.BusinessLogic/Services/Implementations/ClientService.cs]: Move lifetime to constant or env.variable

File: User.BusinessLogic/Services/Implementations/ClientService.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        var newClientModel = newClient.Adapt<ClientModel>();
+
+        return newClientModel;
+    }
+
+    public async Task<ClientModel> UpdateAsync(ClientModel newClientModel, CancellationToken cancellationToken)
+    {
+        var clientToUpdate = await clientRepository.GetByIdAsync(newClientModel.Id, cancellationToken)
+            ?? throw new NotFoundException(ExceptionMessages.NotFound(nameof(UserEntity), newClientModel.Id));
+
```
Comment (5Ka-me) [inline User.BusinessLogic/Services/Implementations/ClientService.cs]: Won't it throw an error in repo by EF?

File: User.BusinessLogic/Services/Implementations/ClientService.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        return clientToReturn;
+    }
+
+    public async Task DeleteAsync(Guid id, CancellationToken cancellationToken)
+    {
+        var clientToDelete = await clientRepository.GetByIdAsync(id, cancellationToken)
+            ?? throw new NotFoundException(ExceptionMessages.NotFound(nameof(UserEntity), id));
+
+        await clientRepository.RemoveAsync(clientToDelete, cancellationToken);
```
Comment (5Ka-me) [inline User.BusinessLogic/Services/Implementations/ClientService.cs]: Wouldn't it better to move this logic to repo?

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

File: User.BusinessLogic/Services/Implementations/ClientService.cs (hunk 3 above)
Comment (DmitriyG0r4anin) [reply User.BusinessLogic/Services/Implementations/ClientService.cs]: will be resolved in specified branch

File: User.BusinessLogic/Extensions/Cache.cs [hunk 4]
```diff
@@ -17,17 +19,19 @@ internal static class Cache
...
     }
 
-    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, TimeSpan cacheLifetime, string key, CancellationToken cancellationToken)
+    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, string key, CancellationToken cancellationToken)
     {
         if (data is null)
             return;
 
+        if (!int.TryParse(lifetime, out var lifetimeMinutes))
+            return;
```
Comment (5Ka-me) [inline User.BusinessLogic/Extensions/Cache.cs]: Don't you think to throw an exception instead of returning null?

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

File: User.BusinessLogic/Extensions/Cache.cs [hunk 5]
```diff
@@ -17,17 +19,19 @@ internal static class Cache
...
 
-    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, TimeSpan cacheLifetime, string key, CancellationToken cancellationToken)
+    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, string key, CancellationToken cancellationToken)
     {
-        if (data is null)
+        if (data is null || lifetime is null)
             return;
 
+        if (!double.TryParse(lifetime, out var lifetimeMinutes))
+            throw new InvalidDataException("lifetime is invalid");
```
Comment (5Ka-me) [inline User.BusinessLogic/Extensions/Cache.cs]: Is the second condition include the first one?

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

File: User.BusinessLogic/Extensions/Cache.cs [hunk 6]
```diff
@@ -17,17 +19,19 @@ internal static class Cache
...
     }
 
-    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, TimeSpan cacheLifetime, string key, CancellationToken cancellationToken)
+    public static async Task CacheData<T>(this IDistributedCache distributedCache, T data, string key, CancellationToken cancellationToken)
     {
         if (data is null)
             return;
 
+        if (!double.TryParse(lifetime, out var lifetimeMinutes))
+            lifetimeMinutes = 3;
```
Comment (5Ka-me) [inline User.BusinessLogic/Extensions/Cache.cs]: Better to move it to the constant

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

**Review by 5Ka-me — APPROVED** [review]

**Pull request merged**